### PR TITLE
Fixes set belongsToMany with scope (v4)

### PR DESCRIPTION
**Pull request by pola88**
see: https://github.com/sequelize/sequelize/pull/7532

### CODE REVIEW

File: test/integration/associations/belongs-to-many.test.js [hunk 1]
```diff
@@ -665,6 +665,60 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
         expect(_tags).to.have.length(1);
       });
     });
+
+    it('updating association via set associations with scope', function() {
+      var ItemTag = this.sequelize.define('ItemTag', {
+                      id : { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
```
Comment (sushantdhiman) [inline test/integration/associations/belongs-to-many.test.js]: Indentation

File: test/integration/associations/belongs-to-many.test.js [hunk 2]
```diff
@@ -665,6 +665,60 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
         expect(_tags).to.have.length(1);
       });
     });
+
+    it('updating association via set associations with scope', function() {
+      var ItemTag = this.sequelize.define('ItemTag', {
```
Comment (sushantdhiman) [inline test/integration/associations/belongs-to-many.test.js]: Use ES6 where possible.

File: test/integration/associations/belongs-to-many.test.js [hunk 3]
```diff
@@ -665,6 +665,60 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
...
+        through: { model: ItemTag, unique: false, scope: { taggable: 'post' } },
+        foreignKey: 'taggable_id'
+      });
+
+      Comment.belongsToMany(Tag, {
+        through: { model: ItemTag, unique: false, scope: { taggable: 'comment' } },
+       foreignKey: 'taggable_id'
+      });
+
+      return this.sequelize.sync({ force: true }).then(function() {
```
Comment (sushantdhiman) [inline test/integration/associations/belongs-to-many.test.js]: Could be a lot simpler with just arrow functions

File: lib/associations/belongs-to-many.js [hunk 4]
```diff
@@ -562,7 +562,7 @@ class BelongsToMany extends Association {
         const where = {};
         where[identifier] = sourceInstance.get(sourceKey);
         where[foreignIdentifier] = obsoleteAssociations.map(obsoleteAssociation => obsoleteAssociation[foreignIdentifier]);
-
+        _.assign(where, association.through.scope);
```
Comment (sushantdhiman) [inline lib/associations/belongs-to-many.js]: This handles destroy case, We also need to handle the other case when associations are updating and a test as well

File: test/integration/associations/belongs-to-many.test.js [hunk 5]
```diff
@@ -665,6 +665,60 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
...
+      }).bind({}).spread(function(post, comment, tag, secondTag) {
+        this.post = post;
+        this.comment = comment;
+        this.tag = tag;
+        this.secondTag = secondTag;
+        return this.post.setTags([this.tag, this.secondTag]);
+      }).then(function() {
+        return this.comment.setTags([this.tag, this.secondTag]);
+      }).then(function() {
+        return this.post.setTags([this.tag]);
```
Comment (sushantdhiman) [inline test/integration/associations/belongs-to-many.test.js]: Confirm if tag count is 1 as well, `this.post.getTags`

**Comment by pola88** [comment]
Thanks @sushantdhiman , I will do the changes that you suggested me

File: lib/associations/belongs-to-many.js (hunk 4 above)
Comment (felixfbecker) [inline lib/associations/belongs-to-many.js]: Use `Object.assign()`

**Comment by felixfbecker** [comment]
LGTM besides usage of ES6.
We should really get on with the switch to ESLint, it can guide with all this

**Comment by janmeier** [comment]
I already ninja fixed it ;)! https://github.com/sequelize/sequelize/blob/master/package.json#L96

**Comment by felixfbecker** [comment]
Let's add all the rules then 💪

**Comment by pola88** [comment]
I'm sorry guys, I was busy, I will try to do the changes that you suggest this weekend

**Comment by pola88** [comment]
Hey guys,
let me know if you see something wrong or to improve,
Thanks.

**Comment by felixfbecker** [comment]
Build is failing

**Comment by pola88** [comment]
@felixfbecker There is one test that is failing, but I think it's not related with my changes.
`"after each" hook for "when we don't have the correct server details"`

**Comment by pola88** [comment]
Hi @sunshinewyin 
I think that I have already tested what you say in another PR that I have made before.

**Comment by blanchma** [comment]
This is just what I was looking for. @sunshinewyin  are you going to merge this PR during this week?

**Comment by pola88** [comment]
Any news about this?
I need this merged

**Comment by blanchma** [comment]
@sushantdhiman @felixfbecker @janmeier I need this PR approved. All the changes were addressed by @pola88

**Comment by sushantdhiman** [comment]
Hi @pola88 , Sorry it took so long to re-review and finally merge this. Thanks for this PR :+1:

**Comment by sushantdhiman** [comment]
Released as `v4.7.2`

**Comment by pola88** [comment]
Thanks @sushantdhiman

**Pull request merged**